plugins/outputs/cloudwatch/cloudwatch.go

```diff
@@ -582,9 +583,16 @@ func BuildDimensions(mTags map[string]string) []*cloudwatch.Dimension {
 		})
 	}
 
+	for key, value := range globalDimensions {
```

Comment:
One thing that I would like to try to optimize is how we introduce the global dimensions. I'm not sure if there is a good way to, but basically what we're doing here is taking a static set of tags and pushing them onto an empty slice for every metric we emit. I want to believe there's a more efficient approach. Haven't thought of how yet

Reply:
I might look at this on another branch (based on this one) to keep this branch a bit more tidy.

Reply:
OK I've had a quick look at this. The best way I can see of doing more of the work up-front for these global dimensions is to put some of the logic into the `startRoutines()` function. In there we could do the work of constructing a slice of dimensions (`*cloudwatch.Dimension[]`) based on the `GlobalDimensions` map before we start doing any actual work.

What this would mean, though, is that the testing becomes a bit funky because either:

- You have to do a bunch of slower (but higher fidelity) tests where you actually start the routines, publish some metrics, and then check that the correct dimensions were used
- Or you have to test the two bits of dimensions processing separately (you could stick the pre-processing of the global dimensions into one function and then pass in the `*cloudwatch.Dimension[]` to the `BuildDimensions` function) and hope that the two actually get used correctly in practice.

If you have a particular preference for one of these testing options and think that this is worth pursuing, please let me know and I'll give it a go.

Reply:
Thanks for looking into it. My initial reaction is that trying to do this pre-optimization is too disruptive. I am _mildly_ concerned with always computing the global dimensions since my understanding is that this happens to every metric. 

I think one thing that I'd like to verify is agent CPU/mem load before and after this change. We could/should test with statsd to have control over the metric throughput and then we can observe CPU and memory usage.